### PR TITLE
Celery 4 support

### PR DESCRIPTION
**Pull request by jlward**

### CODE REVIEW

File: requirements/tests.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-django-nose==1.4.3  # 1.4.4 drops django <1.8 support
+django-nose==1.4.5  # 1.4.4 drops django <1.8 support
```
Comment (jlward) [inline requirements/tests.txt]: Need to remove this comment.

File: tox.ini [hunk 2]
```diff
@@ -1,21 +1,20 @@
 [tox]
 envlist =
-	py27-django{1.6.X,1.7.X,1.8.X,1.9.X}-djangocelery3.1.X,
-	py27-django{1.6.X,1.7.X,1.8.X,1.9.X}-celery3.1.X,
-	py35-django{1.8.X,1.9.X}-djangocelery3.1.X,
-	py35-django{1.8.X,1.9.X}-celery3.1.X,
+	py27-django{1.8.X,1.11.X}-djangocelery3.1.X-celery3.1.X,
+	py27-django{1.8.X,1.11.X}-celery{3.1.X,4.1.X},
+	# py35-django{1.8.X,1.11.X}-djangocelery3.1.X,
```
Comment (jlward) [inline tox.ini]: Should uncomment these lines.

File: jobtastic/tests/test_broker_fallbacks.py [hunk 3]
```diff
@@ -67,33 +62,35 @@ def calculate_result(self, result, **kwargs):
...
         self.old_broker_host = self.app.conf.BROKER_HOST or ''
 
         # Modifying the broken host name simulates the task broker being
         # 'unresponsive'
         # We need to make this modification in 3 places because of version
         # backwards compatibility
-        self._set_broker_host('amqp://')
-        self.app.conf['BROKER_CONNECTION_RETRY'] = False
-        self.app.conf['BROKER_POOL_LIMIT'] = 1
+        self._set_broker_host('foo')
```
Comment (jlward) [inline jobtastic/tests/test_broker_fallbacks.py]: For some reason, in celery4 these tests hang. Which is less than ideal.

File: jobtastic/tests/test_broker_fallbacks.py [hunk 4]
```diff
@@ -67,18 +62,17 @@ def calculate_result(self, result, **kwargs):
...
         os.environ['CELERY_BROKER_URL'] = new_value
-        self.app.conf.BROKER_URL = new_value
-        self.app.conf.BROKER_HOST = new_value
+        self.app.conf['BROKER_URL'] = new_value
+        self.app.conf['BROKER_WRITE_URL'] = new_value
+        self.app.conf['BROKER_READ_URL'] = new_value
 
-    def setup(self):
+    def setUp(self):
         # lowercase on purpose. AppCase calls `self.setup`
```
Comment (jlward) [inline jobtastic/tests/test_broker_fallbacks.py]: This comment should be removed

File: tox.ini [hunk 5]
```diff
@@ -1,21 +1,23 @@
 [tox]
 envlist =
-	py27-django{1.6.X,1.7.X,1.8.X,1.9.X}-djangocelery3.1.X,
-	py27-django{1.6.X,1.7.X,1.8.X,1.9.X}-celery3.1.X,
-	py35-django{1.8.X,1.9.X}-djangocelery3.1.X,
-	py35-django{1.8.X,1.9.X}-celery3.1.X,
+	py27-django{1.8.X,1.11.X}-djangocelery3.1.X-celery3.1.X,
+	py27-django{1.8.X,1.11.X}-celery3.1.X,
+	py27-django{1.8.X,1.11.X}-celery4.1.X-kombu4.0.2,
+	# py35-django{1.8.X,1.11.X}-djangocelery3.1.X,
```
Comment (jlward) [inline tox.ini]: Need to uncomment these guys.

File: tox.ini [hunk 6]
```diff
@@ -1,21 +1,23 @@
...
 	celery3.1.X: celery>=3.1,<3.2
+	celery4.1.X: celery>=4,<5
 	djangocelery3.1.X: django-celery>=3.1,<3.2
-	django1.6.X: django>=1.6,<1.7
-	django1.7.X: django>=1.7,<1.8
 	django1.8.X: django>=1.8,<1.9
-	django1.9.X: django>=1.9,<1.10
+	django1.11.X: django>=1.11,<1.12
+    # The newest version of kombu hangs forever when running tests with celery4
+    kombu4.0.2: kombu==4.0.2
```
Comment (jlward) [inline tox.ini]: Should probably fix the whitespace in here too.

File: jobtastic/task.py [hunk 7]
```diff
@@ -15,8 +15,11 @@
 
 import psutil
 
-from celery.datastructures import ExceptionInfo
-from celery.five import class_property
+from billiard.einfo import ExceptionInfo
+try:  # Celery 3
+    from celery.five import class_property
+except ImportError:  # Celery 4
+    from celery.local import class_property
```
Comment (bocribbz) [inline jobtastic/task.py]: i feel like celery4 should be the default and celery3 the exception.
but not a big deal

File: jobtastic/task.py [hunk 8]
```diff
@@ -157,7 +160,19 @@ def _get_possible_broker_errors_tuple(self):
...
 
-        return dummy_conn.connection_errors + dummy_conn.channel_errors
+        possible_broker_errors = (
+            dummy_conn.connection_errors + dummy_conn.channel_errors
+        )
+        try:
+            from kombu.exceptions import OperationalError
+            # In celery 4.x when a broker is not connected it throws an
+            # OperationalError from kombu. It should also be noted that the
+            # newest version of kombu actually hangs forever. So we need to peg
```
Comment (bocribbz) [inline jobtastic/task.py]: what is the newest version of kombu?

Comment (jlward) [reply jobtastic/task.py]: 4.1.0. We are pegging tests to use 4.0.2

Comment (bocribbz) [reply jobtastic/task.py]: should we put that version in the comment instead of `newest version`?

Comment (bocribbz) [reply jobtastic/task.py]: so whoever works on this next knows what version of kombu was broken from the comment, isntead of digging which one was at the time of this PR

File: jobtastic/tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    if not celery.VERSION[0] == 4:
+        yield
+    else:
+        # In celery 4.x you cannot retrieve the status of a task if it's eager.
+        # This gets around that.
+        with mock.patch(
+            'celery.backends.base.Backend._ensure_not_eager',
+            autospec=True,
+        ):
+            yield
```
Comment (bocribbz) [inline jobtastic/tests/__init__.py]: im not really sure what is going on here.
i mean i understand that it's patching `celery.backends.base.Backend._ensure_not_eager` but what does that have to do with `assert meta.status == PROGRESS`?

File: jobtastic/task.py [hunk 10]
```diff
@@ -157,7 +160,19 @@ def _get_possible_broker_errors_tuple(self):
...
+            # versions of kombu until that gets fixed.
```
Comment (bocribbz) [inline jobtastic/task.py]: where is the kombu version peg happening?

File: jobtastic/tests/__init__.py (hunk 9 above)
Comment (jlward) [reply jobtastic/tests/__init__.py]: The comment above is supposed to be explaining it. But in celery4, you cannot check the status of an eager task. So if we want to check it anyway, we need to skip the check ensuring that the task isn't eager.

File: jobtastic/task.py (hunk 10 above)
Comment (jlward) [reply jobtastic/task.py]: In tox.

Comment (bocribbz) [reply jobtastic/task.py]: shoudnt it be done in requirements too?

File: jobtastic/tests/__init__.py (hunk 9 above)
Comment (bocribbz) [reply jobtastic/tests/__init__.py]: ah so this patching, makes celery think the task is not eager?

File: jobtastic/task.py (hunk 10 above)
Comment (bocribbz) [reply jobtastic/task.py]: the point being that it would save some headache if somebody goes on the newer version of kombu and then installs jobtastic...

Comment (bocribbz) [reply jobtastic/task.py]: i know kombu is not actually a requirement of jobtastic... hmm not sure what the best approach is here.

maybe it should be added to readme/changelog

Comment (bocribbz) [reply jobtastic/task.py]: or maybe im not understanding things right. is the problem only in tests or would it actually cause an app to hang?

File: jobtastic/tests/__init__.py (hunk 9 above)
Comment (jlward) [reply jobtastic/tests/__init__.py]: Pretty much, yea.

File: jobtastic/task.py (hunk 10 above)
Comment (jlward) [reply jobtastic/task.py]: I added a note in the readme.

**Pull request merged**